src/squidpy/gr/_niche.py

```diff
@@ -0,0 +1,432 @@
+
+    deg = np.array(np.sum(adj, axis=1)).squeeze()
+    with np.errstate(divide="ignore"):
+        deg_inv = 1 / deg
+    deg_inv[deg_inv == float("inf")] = 0
+
+    return spdiags(deg_inv, 0, len(deg_inv), len(deg_inv)) * adj
+
+
+def _aggregate(adata: AnnData, normalized_adjacency_matrix: sps.spmatrix, aggregation: str = "mean") -> Any:
+    """aggregate count and adjacency matrix either by mean or variance"""
+    # TODO: add support for other aggregation methods
+    if aggregation == "mean":
+        aggregated_matrix = normalized_adjacency_matrix @ adata.X
+    elif aggregation == "variance":
+        mean_matrix = (normalized_adjacency_matrix @ adata.X).toarray()
+        X_to_arr = adata.X.toarray()
+        mean_squared_matrix = normalized_adjacency_matrix @ (X_to_arr * X_to_arr)
+        aggregated_matrix = mean_squared_matrix - mean_matrix * mean_matrix
+    else:
+        raise ValueError(f"Invalid aggregation method '{aggregation}'. Please choose either 'mean' or 'variance'.")
+
+    return aggregated_matrix
+
+
+def _get_GMM_clusters(A: np.ndarray[np.float64, Any], n_components: int, random_state: int) -> Any:
+    """Returns niche labels generated by GMM clustering.
+    Compared to cellcharter this approach is simplified by using sklearn's GaussianMixture model without stability analysis."""
+
+    gmm = GaussianMixture(n_components=n_components, random_state=random_state, init_params="random_from_data")
+    gmm.fit(A)
+    labels = gmm.predict(A)
+
+    return labels
+
+
+def mean_fide_score(
+    adatas: AnnData | list[AnnData],
+    library_key: str,
+    slide_key: str | None = None,
+    n_classes: int | None = None,
+) -> float:
+    """Mean FIDE score over all slides. A low score indicates a great domain continuity."""
+    return float(
+        np.mean([fide_score(adata, library_key, n_classes=n_classes) for adata in _iter_uid(adatas, slide_key)])
+    )
+
+
+def fide_score(adata: AnnData, library_key: str, n_classes: int | None = None) -> float:
+    """
+    F1-score of intra-domain edges (FIDE). A high score indicates a great domain continuity.
+
+    The F1-score is computed for every class, then all F1-scores are averaged. If some classes
+    are not predicted, the `n_classes` argument allows to pad with zeros before averaging the F1-scores.
+    """
+    i_left, i_right = adata.obsp["spatial_connectivities"].nonzero()
+    classes_left, classes_right = (
+        adata.obs.iloc[i_left][library_key],
+        adata.obs.iloc[i_right][library_key],
+    )
+
+    f1_scores = metrics.f1_score(classes_left, classes_right, average=None)
+
+    if n_classes is None:
+        return float(f1_scores.mean())
+
+    assert n_classes >= len(f1_scores), f"Expected {n_classes:=}, but found {len(f1_scores)}, which is greater"
+
+    return float(np.pad(f1_scores, (0, n_classes - len(f1_scores))).mean())
+
+
+def jensen_shannon_divergence(adatas: AnnData | list[AnnData], library_key: str, slide_key: str | None = None) -> float:
```

Comment:
I think scipy has implemention for this as well.
https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.jensenshannon.html